### ISSUE TITLE
Where to keep message definitions, generated classes and how to version them? - good practices

### ISSUE THREAD

**Issue by kdziurdz** [grpc/grpc-java#5558]
Hi there.

I have question to people who are experienced in using gRPC in multiple microservices architecture written in Java.

TLDR: Where is the single source of truth for message/service definition and how to integrate it with CD?

Do you keep .proto files within repository that provides gRPC server or separated code base?

Do you compile .jar with those classes and pushes it to private repository to make them accessible to other consuming apps or they should generate those classes on their own?

How do you version those packages?

I imagine this situation: Me and my team develop microservice Order. We define our OrderService and I/O messages. We deploy our OrderApp v1.0. and OrderServiceProtos v1.0 (artifact in private repo)

There is another team developing Customer service and they want to use our OrderService to fetch some data. I assume their only resposibility is to download jar with our generated classes, create Channel, create Stub and go fetch some data.

Next week my team add some method to OrderService. We deploy OrderApp v1.1. and OrderServiceProtos v1.1 (artifact in private repo). We notice another team about update. At that moment they know about update and they can fetch the most fresh jar.

So I wonder: ​ Should I bump OrderServiceProtos everytime I deploy new version of OrderApp ?

What if our CD pipeline runs many times everyday (20+)? I don't want to compile and deploy OrderServiceProtos v1.3 that looks exactly the same like v1.2 (because only the App part was updated and CD pipeline ran)

I am wondering if anyone of you had this same considerations.

Or maybe flow with published artifact as a single source of truth is not correct?

**Comment by ST-DDT** [comment]
We use a separate repository for the proto files.
That way we can call the API version with a fixed number and are still able to increment the service/implementation version.
Users can then either use our API.jar directly or use the contained proto files to generate their own variants.
Usually we also include a interface_version file option to make the API version accessible to other platforms.

I recommend using something like Maven's SNAPSHOT versioning. If you are not ready call the version SNAPSHOT, if you want to release it remove the SNAPSHOT and never ever change that version again.
(Push it the companies maven repository).

Your service consumes the API.jar in the same way the other teams will. If you need changes in the API increment the API version and use the new jar. If you change something in your implementation, increment the service version.

For the API version naming I strongly recommend using Semantic versioning. For your service you can use whatever version naming you like, but don't forget to document the API version that it provides.

Protos + Generated Classes -> Order-API_v1.0.jar

Order-API_v1.0.jar + Order-Service Implementation Classes + Dependencies -> Order-Service_v1.0.5-fat.jar

Order-API_v1.0.jar + Customer-API_v2.0.jar + Customer-Service Implementation Classes + Dependencies  -> Customer-Service-20190409-API-v2.0.jar

**Comment by carl-mastrangelo** [comment]
@kdziurdz I'm afraid gRPC team can't answer this question.  We have practices that we use, and they work for us, but it's not of a strong enough conviction to recommend to everyone.   I am going to close this for our own issue management, but others (like @ST-DDT ) can provide their experience.
